### PR TITLE
Metric Server - consolidate apiserver to one port

### PR DESCRIPTION
**Pull request by BojanZelic**
_Provide a description of what has been changed_

### Checklist
- [x] Changelog has been updated and is aligned with our [changelog requirements](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#Changelog)
- [x] A PR is opened to update the documentation on ([repo](https://github.com/kedacore/keda-docs)) https://github.com/kedacore/keda-docs/pull/1225
- [x] Commits are signed with Developer Certificate of Origin (DCO - [learn more](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#developer-certificate-of-origin-signing-your-work))

Fixes [#4460](https://github.com/kedacore/keda/issues/4460)

Relates to https://github.com/kedacore/keda-docs/pull/1225

### CODE REVIEW

File: cmd/adapter/main.go [hunk 1]
```diff
@@ -131,6 +137,40 @@ func (a *Adapter) makeProvider(ctx context.Context) (provider.ExternalMetricsPro
...
+}
+
+func RunMetricsServer() {
+	h := getMetricHandler()
+	mux := http.NewServeMux()
+	mux.Handle("/metrics", h)
+
+	metricsBindAddress := fmt.Sprintf(":%v", metricsAPIServerPort)
+
+	go func() {
```
Comment (zroubalik) [inline cmd/adapter/main.go:164]: Can we please use here the stopCh channel?

File: cmd/adapter/main.go [hunk 2]
```diff
@@ -131,6 +137,40 @@ func (a *Adapter) makeProvider(ctx context.Context) (provider.ExternalMetricsPro
 	return kedaprovider.NewProvider(ctx, logger, mgr.GetClient(), *grpcClient), stopCh, nil
 }
 
+func getMetricHandler() http.HandlerFunc {
```
Comment (zroubalik) [inline cmd/adapter/main.go:140]: could you please document and explain this function in a comment?

File: cmd/adapter/main.go [hunk 3]
```diff
@@ -131,6 +137,40 @@ func (a *Adapter) makeProvider(ctx context.Context) (provider.ExternalMetricsPro
...
+
+	// Return handler that serves metrics from both legacy and controller-runtime registry
+	return func(w http.ResponseWriter, req *http.Request) {
+		legacyregistry.Handler().ServeHTTP(w, req)
+
+		kubemetrics.HandlerFor(ctrlmetrics.Registry, kubemetrics.HandlerOpts{}).ServeHTTP(w, req)
+	}
+}
+
+func RunMetricsServer() {
```
Comment (zroubalik) [inline cmd/adapter/main.go:157]: could you please document and explain this function in a comment?

```
// RunMetricsServer runs...
```

Comment (BojanZelic) [reply cmd/adapter/main.go]: fixed

File: cmd/adapter/main.go (hunk 2 above)
Comment (BojanZelic) [reply cmd/adapter/main.go:141]: fixed

File: cmd/adapter/main.go (hunk 1 above)
Comment (BojanZelic) [reply cmd/adapter/main.go:170]: fixed;

**Comment by JorTurFer** [comment]
This looks amazing! Thanks for the hard research that you have done ❤️ 
Could you extend e2e tests to check these new metrics? There is already a test for other prometheus metrics: https://github.com/kedacore/keda/blob/main/tests/sequential/prometheus_metrics/prometheus_metrics_test.go

File: cmd/adapter/main.go [hunk 4]
```diff
@@ -131,6 +137,57 @@ func (a *Adapter) makeProvider(ctx context.Context) (provider.ExternalMetricsPro
...
+		err := server.ListenAndServe()
+		if err != nil {
+			panic(err)
+		}
+	}()
+
+	go func() {
+		<-stopCh
+		logger.Info("Shutting down the /metrics server gracefully...")
+		ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
```
[the remark this answers is not in the document]
Comment (BojanZelic) [reply cmd/adapter/main.go]: good point, fixed;

**Comment by BojanZelic** [comment]
@JorTurFer added the additional e21e tests to cover both the controller-runtime and apiserver metrics

File: cmd/adapter/main.go [hunk 5]
```diff
@@ -131,6 +136,55 @@ func (a *Adapter) makeProvider(ctx context.Context) (provider.ExternalMetricsPro
...
+
+	// Return handler that serves metrics from both legacy and controller-runtime registry
+	return func(w http.ResponseWriter, req *http.Request) {
+		legacyregistry.Handler().ServeHTTP(w, req)
+
+		kubemetrics.HandlerFor(ctrlmetrics.Registry, kubemetrics.HandlerOpts{}).ServeHTTP(w, req)
+	}
+}
+
+// RunMetricsServer runs a http listener and handles the /metrics endpoint
```
Comment (zroubalik) [inline cmd/adapter/main.go:157]: just a very small nit, could you please add another sentence or two describing why we need this? The metrics consolidation...

Thanks!

File: CHANGELOG.md [hunk 6]
```diff
@@ -54,7 +54,7 @@ To learn more about active deprecations, we recommend checking [GitHub Discussio
 - **Governance**: KEDA transitioned to CNCF Graduated project ([#63](https://github.com/kedacore/governance/issues/63))
 
 ### Improvements
-
+- **General**: Add apiserver prometheus metrics to keda metric server ([#4460](https://github.com/kedacore/keda/issues/4460))
```
Comment (zroubalik) [inline CHANGELOG.md:57]:
```suggestion
- **General**: Add apiserver Prometheus metrics to KEDA Metric Server ([#4460](https://github.com/kedacore/keda/issues/4460))
```

**Review by JorTurFer — APPROVED** [review]
LGTM! ❤️

**Review by zroubalik — APPROVED** [review]

**Pull request merged**